pkg/csi/service/node.go

```diff
@@ -360,3 +452,100 @@ func contains(list []string, item string) bool {
 	}
 	return false
 }
+
+func verifyVolumeAttached(volID string) (string, error) {
+	if volID == "" {
+		return "", status.Error(codes.InvalidArgument,
+			"Volume ID required")
+	}
+
+	// Check that volume is attached
+	volPath, err := getDiskPath(volID, nil)
+	if err != nil {
+		return "", status.Errorf(codes.Internal,
+			"Error trying to read attached disks: %v", err)
+	}
+	if volPath == "" {
+		return "", status.Errorf(codes.NotFound,
+			"Volume ID: %s not attached to node", volID)
+	}
+
+	return volPath, nil
+}
+
+func verifyTargetDir(target string) error {
+	if target == "" {
+		return status.Error(codes.InvalidArgument,
+			"target path required")
+	}
+
+	tgtStat, err := os.Stat(target)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return status.Errorf(codes.FailedPrecondition,
+				"target: %s not pre-created", target)
+		}
+		return status.Errorf(codes.Internal,
+			"failed to stat target, err: %s", err.Error())
+	}
+
+	// This check is mandated by the spec, but this would/should fail if the
+	// volume has a block accessType. Maybe staging isn't intended to be used
+	// with block? That would make sense you cannot share the volume for block.
+	if !tgtStat.IsDir() {
+		return status.Errorf(codes.FailedPrecondition,
+			"existing path: %s is not a directory", target)
+	}
+
+	return nil
+}
+
+// mkdir creates the directory specified by path if needed.
+// return pair is a bool flag of whether dir was created, and an error
+func mkdir(path string) (bool, error) {
+	st, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		if err := os.Mkdir(path, 0750); err != nil {
+			log.WithField("dir", path).WithError(
+				err).Error("Unable to create dir")
+			return false, err
+		}
+		log.WithField("path", path).Debug("created directory")
+		return true, nil
+	}
+	if !st.IsDir() {
```

Comment:
May want to check if `if err != nil { return err }` here, else `st` could be nil.